commit 93ba69aed6308bb677a2876163c1953fe198392b (refs/changes/77/1477/1)
Author: Richard Bradley <richard.bradley@joyent.com>
Date:   2017-02-09T15:03:57+00:00 (2 years, 8 months ago)
    
    MANTA-3100 manta-adm reports wrong server for zone that has changed servers

diff --git a/lib/adm.js b/lib/adm.js
index 9fa11a8..0f6fd73 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -492,7 +492,12 @@ maAdm.prototype.fetchDeployed = function (callback)
 				for (i = 0;
 				    i < self.ma_instances[svcid].length; i++) {
 					instance = self.ma_instances[svcid][i];
-					cns[instance.params.server_uuid] = null;
+					if (self.ma_vms.hasOwnProperty(
+					    instance['uuid'])) {
+						cns[self.ma_vms[
+						    instance['uuid']]
+						    ['server_uuid']] = null;
+					}
 				}
 			}
 
@@ -1365,7 +1370,8 @@ maAdm.prototype.loadInstances = function ()
 		for (i = 0; i < this.ma_instances[svcid].length; i++) {
 			instance = this.ma_instances[svcid][i];
 			metadata = instance['metadata'];
-			server = instance['params']['server_uuid'];
+			server = this.ma_vms.hasOwnProperty(instance['uuid']) ?
+			    this.ma_vms[instance['uuid']]['server_uuid'] : '-';
 			gz = this.ma_gzinfo[server];
 			image = this.ma_vms.hasOwnProperty(instance['uuid']) ?
 			    this.ma_vms[instance['uuid']]['image_uuid'] : '-';
@@ -2399,7 +2405,7 @@ maAdm.prototype.auditZkServers = function ()
 	 */
 	nforeign = 0;
 	jsprim.forEachKey(instancesById, function (zkid, zkinstance) {
-		var uuid, serverUuid, ip, entry;
+		var uuid, ip, entry;
 
 		uuid = zkinstance.uuid;
 		if (!entriesByNum.hasOwnProperty(zkid)) {
@@ -2409,8 +2415,7 @@ maAdm.prototype.auditZkServers = function ()
 			return;
 		}
 
-		serverUuid = zkinstance.params['server_uuid'];
-		if (self.ma_cns[serverUuid] === null) {
+		if (!self.ma_vms.hasOwnProperty(uuid)) {
 			nforeign++;
 			return;
 		}
diff --git a/test/common.js b/test/common.js
index 6513f8d..6da2f95 100644
--- a/test/common.js
+++ b/test/common.js
@@ -220,6 +220,7 @@ function generateFakeBase(fakeDeployed, azCount) {
 			if (masterAz) {
 				fakeBase['vms'][id] = {
 					'image_uuid': params['imgid'],
+					'server_uuid': params['cnid'],
 					'nics': [ {
 						'primary': true,
 						'ip4addr': '0.0.0.0'
diff --git a/test/tst.oneach_exec.js b/test/tst.oneach_exec.js
index b130b89..4a3c325 100644
--- a/test/tst.oneach_exec.js
+++ b/test/tst.oneach_exec.js
@@ -469,6 +469,7 @@ function setupMockManta(_, callback)
 		    'metadata': { 'SHARD': '1', 'DATACENTER': 'test' }
 		});
 		fakeDeployedTopology.vms[zoneid] = {
+		    'server_uuid': cnid,
 		    'nics': [ {
 			'primary': true,
 			'ip': '10.0.0.' + (i + 1)
